File: app/src/main/java/com/chess/cryptobot/view/adapter/RecyclerViewAdapter.java
package com.chess.cryptobot.view.adapter;

import androidx.recyclerview.widget.RecyclerView;

import com.chess.cryptobot.content.ContextHolder;
import com.chess.cryptobot.model.ViewItem;

public abstract class RecyclerViewAdapter<T extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<T> {
    private final ContextHolder holder;

    RecyclerViewAdapter(ContextHolder holder) {
        this.setHasStableIds(true);
        this.holder = holder;
    }

    public String itemNameByPosition(int position) {
        return getItemByPosition(position).getName();
    }

    ViewItem getItemByPosition(int position) {
        return this.holder.getViewItems().get(position);
    }

    public void notifyItemInserted() {
        this.notifyItemInserted(getItemCount());
    }

    public void deleteItem(int position) {
        this.notifyItemRemoved(position);
    }

    public void updateItem(ViewItem item) {
        int index = this.holder.getViewItems().indexOf(item);
        if (index >= 0) {
            this.notifyItemChanged(index);
        }
    }

    @Override
    public int getItemCount() {
        return holder.getViewItems().size();
    }

    @Override
    public long getItemId(int position) {
        ViewItem viewItem = holder.getViewItems().get(position);
        return viewItem.getName().hashCode();
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }
}

File: app/src/main/java/com/chess/cryptobot/view/adapter/RecyclerViewAdapter.kt
package com.chess.cryptobot.view.adapter

import androidx.recyclerview.widget.RecyclerView
import com.chess.cryptobot.content.ContextHolder
import com.chess.cryptobot.model.ViewItem

abstract class RecyclerViewAdapter<T : RecyclerView.ViewHolder?> internal constructor(holder: ContextHolder) : RecyclerView.Adapter<T>() {
    private val holder: ContextHolder
    fun itemNameByPosition(position: Int): String {
        return getItemByPosition(position).name
    }

    fun getItemByPosition(position: Int): ViewItem {
        return holder.viewItems[position]
    }

    fun notifyItemInserted() {
        this.notifyItemInserted(itemCount)
    }

    fun deleteItem(position: Int) {
        notifyItemRemoved(position)
    }

    fun updateItem(item: ViewItem?) {
        val index = holder.viewItems.indexOf(item)
        if (index >= 0) {
            this.notifyItemChanged(index)
        }
    }

    override fun getItemCount(): Int {
        return holder.viewItems.size
    }

    override fun getItemId(position: Int): Long {
        val viewItem = holder.viewItems[position]
        return viewItem.name.hashCode().toLong()
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    init {
        enableStableIds()
        this.holder = holder
    }

    private fun enableStableIds() {
        setHasStableIds(true)
    }
}